src/PackagePrivate/MediaWikiNormalizedTermCleaner.php

```diff
@@ -0,0 +1,271 @@
+	private function cleanTermInLangIds( array $termInLangIds ) {
+		if ( $termInLangIds === [] ) {
+			return;
+		}
+
+		$this->logger->debug(
+			'{method}: deleting {count} term_in_lang rows',
+			[
+				'method' => __METHOD__,
+				'count' => count( $termInLangIds ),
+			]
+		);
+
+		$potentiallyUnusedTextInLangIds = $this->selectFieldValuesForPrimaryKey(
+			'wbt_term_in_lang',
+			'wbtl_text_in_lang_id',
+			'wbtl_id',
+			$termInLangIds,
+			__METHOD__
+		);
+
+		$this->dbw->delete(
+			'wbt_term_in_lang',
+			[ 'wbtl_id' => $termInLangIds ],
+			__METHOD__
+		);
+
+		$stillUsedTextInLangIds = $this->dbw->selectFieldValues(
+			'wbt_term_in_lang',
+			'wbtl_text_in_lang_id',
+			[ 'wbtl_text_in_lang_id' => $potentiallyUnusedTextInLangIds ],
+			__METHOD__,
+			[
+				/**
+				 * If we try to clean up a text_in_lang whose last use in a term_in_lang we just
+				 * removed, and simultaneously another request adds a new term_in_lang using that
+				 * text_in_lang, we want one of the following to happen:
+				 *
+				 * 1. Our transaction completes first and removes the text_in_lang. The concurrent
+				 *    request blocks until we’re done, then sees that the text_in_lang is gone and
+				 *    creates it (again) as part of inserting the term_in_lang.
+				 * 2. The other transaction completes first and registers another term_in_lang using
+				 *    that text_in_lang. We block until that’s done and then notice the text_in_lang
+				 *    is still used and don’t clean it up.
+				 *
+				 * For this to work, we need to use 'FOR UPDATE' when checking whether a
+				 * text_in_lang is still used in a term_in_lang, and the other request needs to
+				 * ensure during or after insert of the new term_in_lang that the text_in_lang still
+				 * exists, or create it otherwise. This way, either our check here or the other
+				 * request’s insert will block and wait for the other to complete.
+				 */
+				'FOR UPDATE',
+				// 'DISTINCT', // not supported in combination with FOR UPDATE on some DB types
+			]
+		);
+		$unusedTextInLangIds = array_diff(
+			$potentiallyUnusedTextInLangIds,
+			$stillUsedTextInLangIds
+		);
+
+		$this->cleanTextInLangIds( $unusedTextInLangIds );
+	}
+
+	/**
+	 * Delete the specified text_in_lang rows from the database,
+	 * as well as any text rows that are now unused.
+	 *
+	 * @param int[] $textInLangIds
+	 */
+	private function cleanTextInLangIds( array $textInLangIds ) {
+		if ( $textInLangIds === [] ) {
+			return;
+		}
+
+		$this->logger->debug(
+			'{method}: deleting {count} text_in_lang rows',
+			[
+				'method' => __METHOD__,
+				'count' => count( $textInLangIds ),
+			]
+		);
+
+		$potentiallyUnusedTextIds = $this->selectFieldValuesForPrimaryKey(
+			'wbt_text_in_lang',
+			'wbxl_text_id',
+			'wbxl_id',
+			$textInLangIds,
+			__METHOD__
+		);
+
+		$this->dbw->delete(
+			'wbt_text_in_lang',
+			[ 'wbxl_id' => $textInLangIds ],
+			__METHOD__
+		);
+
+		$stillUsedTextIds = $this->dbw->selectFieldValues(
```

Comment:
This will throw an `InvalidArgumentException` if `$potentiallyUnusedTextIds` is an empty array (which is an extreme(?) edge case, that would only happen if `wbt_text_in_lang` is inconsistent AFAIS). First I feared that MW might `SELECT` without `WHERE` in that case, but it luckily doesn't.

Reply:
Yeah, that shouldn’t ever happen™, and an `InvalidArgumentException` should be fine in that case. (I’d thought about adding an assertion to `selectFieldValuesForPrimaryKey` that the rows returned by master have the right count, but decided against it.)